google/cloud/pubsublite/internal/service.h

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_SERVICE_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUBLITE_INTERNAL_SERVICE_H
+
+#include "google/cloud/future.h"
+#include "google/cloud/status.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+/**
+ * A `Service` object is one that has a `Start` method that must be called
+ * exactly once in order for other functionality of the object to be invoked and
+ * a `Shutdown` method that must be called exactly once when the object's
+ * functionality is no longer needed/before destruction.
+ */
+class Service {
+ protected:
+  Service() = default;
+
+ public:
+  virtual ~Service() = default;
+  Service(const Service&) = delete;
+  Service& operator=(const Service&) = delete;
+  Service(Service&&) = delete;
+  Service& operator=(Service&&) = delete;
+
+  /**
+   * Starts the lifecycle of the object.
+   *
+   * The future returned by this function is satisfied when the object
+   * encounters an error during one of its operations or when `Shutdown` is
+   * called on it. The value the future is satisfied with details what happened
+   * to the object, i.e., `Shutdown` was called, encountered a permanent error,
+   * etc.
+   *
+   * Must be called before any other method and may only be called once.
```

Comment:
The only other method is `Shutdown()`?  Does this requirement apply to member functions added in derived classes?  That seems odd (usually the base class / interface cannot impose restrictions on how the children behave).  If this is the case, then please say that this also applies to the children classes.

Reply:
yup applies to derived classes, added a note.